### ISSUE TITLE
We don't have a TEI-XML version

### ISSUE THREAD

**Issue by normansimonr** [byztxt/byzantine-majority-text#12]
[TEI-XML](http://epapers.bham.ac.uk/1727/5/IGNTP_XML_guidelines_1-4.pdf) is the standard format for transcribing the Greek New Testament. All transcriptions from the [Münster Institute for New Testament Textual Research](http://ntvmr.uni-muenster.de/manuscript-workspace) are made using this format. This format is also used by the acclaimed [CollateX software](https://collatex.net/doc/#xml-input) to automaticaly generate manuscript collations.

I think we should write a converter to make TEI-XML versions of all the books in the New Testament. This could help the Byzantine text gain more traction amongst textual scholars. The XML should of course include all variants within the Byzantine text and against the NA text (basically everything that is in the marginal and lower apparata.)

This could be a good first issue for new collaborators (after we adjust and merge #6.)

**Comment by emg** [comment]
Thanks for the suggestion, @normansimonr – I heartily agree.

What would you see as desirables features a TEI XML conversion?

Would you like to write the converter?

**Comment by countingpine** [comment]
I'm not familiar with the format, but at a glance it looks quite open ended. Does the beta code contain all the information we would want for the XML?

I wonder if it would be worth doing a manual conversion of something short like Philemon, so we have a reference to work towards?

**Comment by emg** [comment]
The TEI format is indeed open-ended. The TEI spec is very, very big.

The advantage of TEI over many other formats include:

- Everything is well documented, including the intended meaning of all elements and attributes.
- That means, there is an authoritative reference to refer to, concerning which elements and which attributes to use.
- This is useful both for people who encode the text and for people who consume the encoded text.
- TEI is a widely used, well-respected standard within the Digital Humanities community.

Münster uses TEI, and so does the German Bible Society for some of its other texts. So TEI is getting some traction within the Bible community.

@countingpine, You ask whether the beta code contains all the information we would want to include in the XML.

I would say no. To serve the intended audicence the best, we would need to include the information contained in the parsed version as well:

(PARSED): https://github.com/byztxt/byzantine-majority-text/tree/master/parsed

Perhaps other information would be useful to add. I don't know exaclty what – but ideas from those who have more experience with the Münster texts would be welcome.

As for how to do it, I agree: It is good, as a first step, to make a target version by hand. This is essentially a way of prototyping the output / designing the output / validating the usefulness of the output, before writing the conversion. @countingpine  – this is an excellent suggestion. Thank you.

Making a prototype by hand also gives the opportunity to reflect on how to make the output most useful.

Perhaps one could even beta-test the prototype, for example, by uploading to some tool. Or by asking for comments from real users. 

As for how to integrate the parsed version, here is how I would go about it:

1. Read the beta version into memory. Call this version (BETA). Make sure the reading is (more or less) correct before proceding.
2. Convert (BETA) to TEI XML. Call this version (TEI-XML). Emit this as TEI XML. Make sure it is right before proceding.
3. Integrate (TEI-XML) with (PARSED), producing a new version of the TEI XML. Call this (TEI-XML-PARSED). Dump this as TEI XML. Make sure this is right before proceding.

This is essentially a series of transformations, with integrity checks along the way to ensure that everything is done right.

I usually find that converting one version of a text to another requires several steps of transformation, even within each of the steps outlined above.

For example, in order to convert the BETA version to TEI XML, several steps may be required before the goal is reached of representing the text as TEI XML. One step would be reading the BETA version into memory. This would probably require interpreting the BETA version with respect to all of the various kinds of information available – for example, the apparatus. Another step might be 

I usually find that outputting the results of each transformation step is a good and necessary way to check the integrity of the conversion.

All steps should be done programmatically, of course, so as to be reproducible. No direct editing by hand should ideally be allowed in any step. The reason for this requirement is that any of the upstream versions are likely to change. For example, Dr. Robinson may come out with a new version of the BETA code. Or he may change a parsing in the PARSED version. Then it is necessary to re-run the conversion.

Also, all steps should include integrity checks to ensure that no information is lost, and that the new representation accurately and usefully reflects the underlying reality.

This means that programmatic integrity checks are usually necessary.

Even the step of reading the input text into memory may require a lot of effort to get right. There are so many things one tends to assume about the structure and content of a text which, upon inspection, fail to meet the test of a reality check. I think we have already discovered how difficult it is to get the Unicode conversion right. 

In other words: This text is important. Making sure it is represented correctly and usefully is part of the task of handling the Word of Truth correctly, akin to the task of being a manuscript scribe in times of yore.

I am so grateful to all of you, who care about this work enough that you would come along and spend time and effort to get these things right. I trust that you will find that the work is rewarding in its own right. But I also trust that you will be given plentiful grace along the journey, and that all of your needs will be met.

Thanks for your collaboration.

Ulrik

**Comment by normansimonr** [comment]
@emg @countingpine Hey guys, I am so sorry for my lateness in replying. I am mostly settled in my new home (new city) and can finally make the time to work on this.

The Münster Institute [has](http://ntvmr.uni-muenster.de/liste?docID=1010001) TEI XML files for a selection of passages from the Robinson-Pierpoint edition. [Here](https://ntvmr.uni-muenster.de/community/vmr/api/transcript/get/?docID=1010001&pageID=ALL&format=teiraw) is what they look like. What I have in mind is something like this, but for the entire New Testament. Our version should also include variants (both those within the Byzantine tradition and the differences between the Byzantine and the ecclectic text.)

I think we do have all the data we need in the CSV version. The only problem is that the CSV version has accents and breathings, which is not standard in the Münster transcriptions. So we need to either

* find a way to convert the raw unaccented BETA code into Unicode and then merge it with the variants info, or
* convert the Unicode CSV files into XML and then find a way to convert the accented text into non-accented text

The second option is attractive because we can then have two versions. An set of unaccented XML files and another one with accented text.

**Comment by countingpine** [comment]
Hi,

From what I'm reading, we have differing opinions on what should go into the XML. @normansimonr, you think we need only  include the information in the XML, but @emg, you want to include the parsed information.

Regarding the accents and punctuation: I think it would be relatively easy to do that, if desired, by simply stripping out all the accent and punctuation characters during the CSV conversion, perhaps into a separate folder. (I think `()=+^|/\`, and `,.;:` cover everything?)

Some care might just be needed to handle text in quotes.  I've suggested a patch in #13 ([countingpine@quoted](https://github.com/countingpine/byzantine-majority-text/commit/quoted)) that is good enough to handle all the text that currently appears in quotes.  (I'm happy to do a PR when the existing one is merged.)

If we want the parsed information and the Greek, it might be possible to substitute the asciified words for the original Greek ones from the no-variants CSV, if they match up perfectly.  I've not tried, but I could probably manage at least an inelegant solution.

**Comment by normansimonr** [comment]
> From what I'm reading, we have differing opinions on what should go into the XML. @normansimonr, you think we need only include the information in the XML, but @emg, you want to include the parsed information.

Oh ok, I forgot about that bit. Including the parsing information would be ideal. I believe we can put it inside `<note>` tags.

> Regarding the accents and punctuation: I think it would be relatively easy to do that, if desired, by simply stripping out all the accent and punctuation characters during the CSV conversion, perhaps into a separate folder. (I think ()=+^|/\, and ,.;: cover everything?)

Makes sense. We can then have accented and unaccented CSVs, and work off of them to make the XML.

> If we want the parsed information and the Greek, it might be possible to substitute the asciified words for the original Greek ones from the no-variants CSV, if they match up perfectly. I've not tried, but I could probably manage at least an inelegant solution.

Sounds good. Part of the problem is that for any given book, Professor Robinson doesn't give us a single source file from which we can derive all the other files. We have three source files for each book ([parsed](https://github.com/byztxt/byzantine-majority-text/tree/master/parsed), [textonly-beta-code](https://github.com/byztxt/byzantine-majority-text/tree/master/textonly-beta-code), and [textonly-online-bible](https://github.com/byztxt/byzantine-majority-text/tree/master/textonly-online-bible).)

**Comment by normansimonr** [comment]
Actually, we could probably "merge" all the three types of the source files in (Unicode) XML files that can then be used as a basis to translate to other formats like CSV and LaTeX, and even Markdown (and by extension Word and HTML using [`pandoc`](https://pandoc.org/).) I might be getting ahead of myself of course, but in any case the first step would be to have the XMLs.

For the CSV converter, this would mean that instead of taking a subset of the source files and converting them directly to CSV, we would first convert to XML and then write a new XML-to-CSV converter.

**Comment by normansimonr** [comment]
I'm closing this issue because we added TEI-XML files in 9a24722de71a4ea3965d9d3318a9f3ce7374fba5
